### PR TITLE
Add code scanner tool

### PR DESCRIPTION
**Pull request by celinval**
This is just a utility tool that allow us to scan a crate and extract some metrics out of it. I also added a script to scan the standard library.

Currently, the tool will produce a few CSV files with raw data.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by tautschnig** [comment]
The missing license information needs to be fixed, but also I am wondering about the version downgrades to some of our dependencies?!

**Comment by celinval** [comment]
Yeah, I need to fix that. Thanks

File: scripts/std-analysis.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+results=/tmp/std_lib_analysis/results
+mkdir $results
+find /tmp/std_lib_analysis/target -name "*.csv" -exec mv {} $results \;
+
+# Create a summary table
+summary=$results/summary.csv
+
+# write header
+echo -n "crate," > $summary
+tr -d [:digit:], < $results/alloc_scan_overall.csv \
```
Comment (qinheping) [inline scripts/std-analysis.sh:97]:
```suggestion
tr -d "[:digit:],;" < $results/alloc_scan_overall.csv \
```
The header now contains leftover "l", i.e., 
`crate,total_fns;,has_loops;,has_for_loops;,has_non_array_for_loops;,no_loops;,has_iterators;,`

File: scripts/std-analysis.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    | tr -s '\n' ',' >> $summary
+echo "" >> $summary
+
+# write body
+for f in $results/*overall.csv; do
+    # Join all crate summaries into one table
+    fname=$(basename $f)
+    crate=${fname%_scan_overall.csv}
+    echo -n "$crate," >> $summary
+    tr -d [:alpha:]_, < $f | tr -s '\n' ',' \
```
Comment (qinheping) [inline scripts/std-analysis.sh:107]:
```suggestion
    tr -d [:alpha:]_,; < $f | tr -s '\n' ',' \
```

File: tools/scanner/src/analysis.rs [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+
+//! Provide different static analysis to be performed in the crate under compilation
+
+use crate::info;
+use csv::WriterBuilder;
+use serde::{ser::SerializeStruct, Serialize, Serializer};
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::visit::{Location, PlaceContext, PlaceRef};
+use stable_mir::mir::{
+    Body, Constant, MirVisitor, Mutability, ProjectionElem, Safety, Terminator, TerminatorKind,
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:12]:
```suggestion
    Body, MirVisitor, Mutability, ProjectionElem, Safety, Terminator, TerminatorKind,
```
Toolchain changes.

File: tools/scanner/src/analysis.rs [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+};
+use stable_mir::ty::{AdtDef, AdtKind, FnDef, GenericArgs, RigidTy, Ty, TyKind};
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:14]:
```suggestion
use stable_mir::ty::{AdtDef, AdtKind, FnDef, GenericArgs, MirConst, RigidTy, Ty, TyKind};
```
Toolchain changes.

File: tools/scanner/src/analysis.rs [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+        let all_items = stable_mir::all_local_items();
+        let (unsafe_fns, safe_fns) = all_items
+            .into_iter()
+            .filter_map(|item| {
+                let kind = item.ty().kind();
+                if !kind.is_fn() {
+                    return None;
+                };
+                let fn_sig = kind.fn_sig().unwrap();
+                let is_unsafe = fn_sig.skip_binder().unsafety == Safety::Unsafe;
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:97]:
```suggestion
                let is_unsafe = fn_sig.skip_binder().safety == Safety::Unsafe;
```
Toolchain changes.

File: tools/scanner/src/analysis.rs [hunk 6]
```diff
@@ -0,0 +1,629 @@
...
+        let (has_unsafe, no_unsafe) = all_items
+            .into_iter()
+            .filter_map(|item| {
+                let kind = item.ty().kind();
+                if !kind.is_fn() {
+                    return None;
+                };
+                let unsafe_ops = FnUnsafeOperations::new(item.name()).collect(&item.body());
+                let fn_sig = kind.fn_sig().unwrap();
+                let is_unsafe = fn_sig.skip_binder().unsafety == Safety::Unsafe;
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:142]:
```suggestion
                let is_unsafe = fn_sig.skip_binder().safety == Safety::Unsafe;
```
Toolchain changes.

File: tools/scanner/src/analysis.rs [hunk 7]
```diff
@@ -0,0 +1,629 @@
...
+    props: FnUnsafeOperations,
+    body: &'a Body,
+}
+
+impl<'a> MirVisitor for BodyVisitor<'a> {
+    fn visit_terminator(&mut self, term: &Terminator, location: Location) {
+        match &term.kind {
+            TerminatorKind::Call { func, .. } => {
+                let fn_sig = func.ty(self.body.locals()).unwrap().kind().fn_sig().unwrap();
+                if fn_sig.value.unsafety == Safety::Unsafe {
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:390]:
```suggestion
                if fn_sig.value.safety == Safety::Unsafe {
```

File: tools/scanner/src/analysis.rs [hunk 8]
```diff
@@ -0,0 +1,629 @@
...
+            }
+            ty.super_visit(self)
+        }
+    }
+}
+
+fn dump_csv<T: Serialize>(mut out_path: PathBuf, data: &[T]) {
+    out_path.set_extension("csv");
+    info(format!("Write file: {out_path:?}"));
+    let mut writer = WriterBuilder::new().delimiter(b';').from_path(&out_path).unwrap();
```
Comment (qinheping) [inline tools/scanner/src/analysis.rs:342]: Is there any specific reason to use delimiter `;` instead of comma `,`?

**Review by qinheping — APPROVED** [review]

File: tools/scanner/src/analysis.rs (hunk 8 above)
Comment (celinval) [reply tools/scanner/src/analysis.rs:342]: Yes, because the name of the functions often contain `,` to separate generic parameters.

**Pull request merged**